src/System.Private.CoreLib/shared/System/Globalization/CompareInfo.cs

```diff
@@ -39,6 +39,9 @@ public partial class CompareInfo : IDeserializationCallback
             ~(CompareOptions.IgnoreCase | CompareOptions.IgnoreSymbols | CompareOptions.IgnoreNonSpace |
               CompareOptions.IgnoreWidth | CompareOptions.IgnoreKanaType | CompareOptions.StringSort);
 
+        [NonSerialized]
```

Comment:
This lock is only used in the Unix-specific file. It should not be in the common file.

Reply:
How can I keep it in the Unix-specific file and ensure that it's going to get initialized before the `Invariant` field?

```log
Unhandled exception. System.TypeInitializationException: The type initializer for 'System.Globalization.CompareInfo' threw an exception. ---> System.ArgumentNullException: Value cannot be null.
   at System.Threading.Monitor.ReliableEnter(Object obj, Boolean& lockTaken)
   at System.Globalization.CompareInfo.InitSort(CultureInfo culture)
   at System.Globalization.CultureInfo.get_CompareInfo()
   at System.Globalization.CompareInfo..cctor()
```

Reply:
Create small inner helper type `private static class SortHandleCache` with just the one `Dictionary<string,IntPtr>` static field.

Once you do that, you also do not need the separate `lock`. You can just lock on the Dictionary.

Reply:
thanks!

> Once you do that, you also do not need the separate lock object. You can just lock on the Dictionary.

I was wondering why in other types we always have a dictionary and a separate object for lock. Does it mean that if I can provide a good encapsulation I should not be having a separate object for lock?

Reply:
It depends on the situation. If you have a convenient object to lock on, there is no reason to create a separate lock object.